### PR TITLE
Prevent FluxBufferTimeout requests < 0 when enough outstanding

### LINKED ISSUE #2839: Error thrown when upstream publisher and subscriber are slow
## Expected Behavior
When items are received from upstream after a buffer has timed out when the downstream subscriber is currently satisfied and cannot accept more data the items should be buffered until the subscriber requests more items.

## Actual Behavior
An error signal is raised ("Could not emit buffer due to lack of requests"). In addition sometimes under the similar circumstances it can request a negative amount of data from the upstream. This happens when the subscriber requests a few batches, gets a few single item lists back after timeouts and then requests another (at this point requested * batchSize can be less that outstanding). Change the flatMap concurrency parameter to 3/4 in the below test to see that happen.

## Steps to Reproduce

```java
@Slf4j
public class BufferTest
{
    public static void main(String[] args)
    {
        Flux<Integer> source = Flux.create(BufferTest::slowEmitter)
                                   .log("source");

        FluxBufferTimeout<Integer, List<Integer>> subject = new FluxBufferTimeout<>(source,
                                                                                    3,
                                                                                    50, TimeUnit.MILLISECONDS,
                                                                                    Schedulers.parallel(),
                                                                                    ArrayList::new);

        subject.log("downstream", Level.INFO, SignalType.REQUEST, SignalType.ON_NEXT)
               .flatMap(list -> Mono.delay(Duration.ofMillis(500)).thenReturn(list), 1)
               .doOnNext(list -> log.info("Output: {}", list))
               .blockLast();
    }

    private static void slowEmitter(final FluxSink<Integer> sink)
    {
        sink.onRequest(amount -> onRequestSlow(sink, amount));
    }

    private static void onRequestSlow(final FluxSink<Integer> sink, final long amount)
    {
        Flux.range(0, (int)amount)
            .delayElements(Duration.ofMillis(500))
            .doOnNext(val -> sink.next(COUNTER.getAndIncrement()))
            .subscribe();
    }
}
```

## Possible Solution
Instead of raising the error leave the data in the buffer and do another flush at the point of the next request from downstream. I have come up with a suggested PR to fix this, will submit if you think that's the right solution.

## Your Environment

* Reactor version(s) used: Spotted in reactor 3.4.6, still present in 3.4.12
* Other relevant libraries versions (eg. `netty`, ...):
* JVM version (`java -version`): openjdk version "11.0.11" 2021-04-20
* OS and version (eg `uname -a`): Linux dev 5.8.0-59-generic #66~20.04.1-Ubuntu SMP Thu Jun 17 11:14:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by simonbasle**
This commit puts a simple guard against negative request in the
BufferTimeoutSubscriber#requestMore method in order to cover cases
where discrepancies are introduced by the timeouts.

Request accounting is not perfect when both a size and a duration can
be specified, given that requests upstream are multiplicated in order
to get a chance at filling buffers.

Hence, this change puts a simple guard which ignores negative requests.

Fixes #2839.

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/publisher/FluxBufferTimeout.java [hunk 1]
```diff
@@ -320,6 +320,9 @@ public void request(long n) {
 		}
 
 		final void requestMore(long n) {
+			if (n < 1L) {
```
Comment (OlegDokuka) [inline reactor-core/src/main/java/reactor/core/publisher/FluxBufferTimeout.java:323]: Sounds like a bug somewhere in the core logic

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxBufferTimeout.java:323]: I'm not entirely sure... 
I think this scenario is valid and leads to the above state:
  - buffer size 10
  - let's assume a first request of 2 buffers
  - let's also assume the source is slow to emit at first and delivers 2 elements, then times out => first buffer request is met with timeout
  - then it emits one more element and stalls again => second buffer request is met with timeout also
  - `request(1)` comes in from downstream

At this point we have `outstanding` at `2x10 - 3` = `17`.
A request for one buffer translates to `requestLimit = 10`.
We have sufficient outstanding request to cover for that buffer and in `requestMore(n)` where `n = requestLimit - outstanding`, `n` is negative.

I think that this check should probably be better made in the `else` branch of `request` though, wdyt ?

**Comment by hepin1989** [comment]
Is there any ETA of this release?

**Comment by simonbasle** [comment]
> Is there any ETA of this release?

yes, next release will be second tuesday of February (assuming this gets approved and merged)

**Review by OlegDokuka — APPROVED** [review]

**Pull request merged**